test/loading.jl

```diff
@@ -2,22 +2,51 @@
 
 using Base.Test
 
-@test @__LINE__ == 5
+# Tests for @__LINE__ inside and outside of macros
+@test (@__LINE__) == 6
+
+macro macro_caller_lineno()
+    @test 9 == (@__LINE__) != __source__.line > 12
+    return __source__.line
+end
+
+@test @macro_caller_lineno() == (@__LINE__) > 12
+
+# @__LINE__ in a macro expands to the location of the macrocall in the source
+# while __source__.line is the location of the macro caller
+macro nested_LINE_expansion()
+    return quote
+        return (@emit_LINE, $(__source__.line))
+    end
+end
+macro nested_LINE_expansion2()
+    return :((@emit_LINE, $(__source__.line)))
+end
+macro emit_LINE()
+    return quote
+        (@__LINE__, $(__source__.line))
+    end
+end
+@test (@emit_LINE) == ((@__LINE__) - 3, @__LINE__)
+@test @nested_LINE_expansion() == ((@__LINE__() - 4, @__LINE__() - 12), @__LINE__())
+@test @nested_LINE_expansion2() == ((@__LINE__() - 5, @__LINE__() - 9), @__LINE__())
```

Comment:
Hah, wouldn't have been a problem with #20895 :-P

How could a deprecation be made to work? I assume this broke packages which were doing moderately "interesting" AST manipulations. Those would just be manipulating the surface syntax via an `Expr` which happened to have `head==:macrocall` so there's nowhere to insert any kind of depwarn, is there?

I could imagine adding it into some kind of AST pattern matching tool inside Compat, and encouraging people to use that when interacting with ASTs containing macros? Could be a lot easier on users than the minimal flag in https://github.com/JuliaLang/Compat.jl/pull/355/files